File: 2116-check-if-a-parentheses-string-can-be-valid/2116-check-if-a-parentheses-string-can-be-valid.cpp
class Solution {
public:
    bool canBeValid(string s, string locked) {
        int n = s.length();
        if(n%2==1){
            return false;
        }
        stack<int> lockStack, st;
        for(int i=0;i<n;i++){
            if(locked[i]=='0'){
                lockStack.push(i);
            }else if(s[i]=='('){
                st.push(i);
            }else{
                if(!st.empty()){
                    st.pop();
                }else if(!lockStack.empty()){
                    lockStack.pop();
                }else{
                    return false;
                } 
            }
        }
        while(!st.empty() && !lockStack.empty() && st.top()<lockStack.top()){
            st.pop();
            lockStack.pop();
        }
        return st.empty();
    }
};
// 

File: 2116-check-if-a-parentheses-string-can-be-valid/2116-check-if-a-parentheses-string-can-be-valid.java
class Solution {
    public boolean canBeValid(String s, String locked) {
        int n = s.length(),i=0;
        if(n%2!=0) return false;
        Stack<Integer> st = new Stack<>();
        Stack<Integer> lc = new Stack<>();
        while(i<n){
            char sc = s.charAt(i), l = locked.charAt(i);
            if(l=='0'){
                lc.push(i);
            }else if(sc=='('){
                st.push(i);
            }else{
                if(!st.isEmpty()){
                    st.pop();
                }else if(!lc.isEmpty()){
                    lc.pop();
                }else{
                    return false;
                }
            }
            i++;
        }
        while(!st.isEmpty() && !lc.isEmpty() && st.peek()<lc.peek()){
            st.pop();
            lc.pop();
        }
        return st.isEmpty();
    }
}